### ISSUE TITLE
m11600 unique salt optimization

### ISSUE THREAD

**Issue by cxxcode** [hashcat/hashcat#3500]
Hello.
I found incorrectly working the module 11600 with several tasks.

Suppose we have 2 tasks:
`$7z$0$19$0$$8$424e43f4da82743b0000000000000000$331058520$16$5$d3887464a99fa0b591bb733ff60891e8`
`$7z$0$19$0$$8$8ea89008420a42a80000000000000000$2326977762$16$5$b2aced7efef389de074fc7394723296f`

 Hashcat run command:
`hashcat.exe -D 1 -m 11600 -a 3 data\tasks.txt ?a?a?a?a`

Hashcat takes them into processing and reports this:
`Hashes: 2 digests; 2 unique digests, 2 unique salts`

Wait, what? 
Salts are empty in all tasks (fields `$0$$`), so they expand to `000000000000` (8 zero bytes) by default, i.e **salts are the same for all tasks**. However, **hascat says about 2 unique salts**.

Let see code in `modeule_hash_decode` function.

Set `seven_zip->salt_buf` from `salt_buf` is correct. [copy_salt_buf__ok](https://github.com/hashcat/hashcat/blob/master/src/modules/module_11600.c#L634)

Set `salt->salt_buf` is wrong, because we write **cipher_text** to the `salt->salt_buf`. [copy_cipher_text_to_salt__error!!!](https://github.com/hashcat/hashcat/blob/master/src/modules/module_11600.c#L695-L698)

This is the reason why hashcat informs about not a single salt.

The correct code should be like this:
`salt->salt_buf[0] = seven_zip->salt_buf[0];`
`salt->salt_buf[1] = seven_zip->salt_buf[1];`
`salt->salt_buf[2] = seven_zip->salt_buf[2];`
`salt->salt_buf[3] = seven_zip->salt_buf[3];`

By the way, the working time in this case will be reduced in proportion to the number of tasks.

Thanks.

**Comment by jsteube** [comment]
Correct. The design of the 7Zip kernel should allow in this case to be set to one unique salt. However, the hook function it is where the decryption is done in the 7Zip kernel. So we can't just change the code in the decoder it would only crack the first entries (after internal sorting), but in theory it's possible.

In general, hooks are designed to have exactly one input and one output. If we want to make this possible, the hook code needs to be rewritten to accept one input but multiple output. And later, the _comp kernel also need to understand that there's multiple hook outputs to be tested. 

If you want to try, there are some plugins that are similar in terms of single input and multiple output. It's just that none of them use a hook. Your goal would be to rewrite the plugin to make use of OPTS_TYPE_DEEP_COMP_KERNEL flag. Best thing to start is to read the section about this in the developer guide. Later you can use for example mode 9500 to see how the logic in _comp kernel needs to be adjusted in order to make use of OPTS_TYPE_DEEP_COMP_KERNEL. The only thing where no reference implementation exists yet is how to transport multiple entries back through the hook channel, but I assume you can just trick and modify hook_success to hook_success[] with a fixed maximum number of entries.

**Comment by philsmd** [comment]
Yeah, I agree too and it is still crazy why they (the 7z SDK, archiving tools) don't use a random salt for each and every new 7z file. This is just ridiculous/amazing and there is probably no good reason for it. Indeed, you could technically simply use something like a "rainbow table" attack for this particular algorithm when you have to look up many such hashes and have a fixed/great word list (a mapping between pass and keys, even through for most other tools/agorithms this makes no sense nowadays, but this algo is special).
